sns/parity_math.c

```diff
@@ -1694,37 +1801,163 @@ static inline uint32_t recov_mat_col(const struct m0_sns_ir_block *alive_block,
+	uint32_t		i;
+	uint32_t		j;
+	uint8_t		       *buf_data;
+	uint32_t		seg_size;
+
+	M0_ENTRY("bvec=%p, buf=%p, count=%u, dir=%s",
+		 bvec, buf, count, (dir?"TRUE":"FALSE"));
+
+	for (j = 0, i = 0; j < count; ++j, i = 0) {
+		m0_bufvec_cursor_init(&cursor, &bvec[j]);
+		buf_data = (uint8_t *)buf[j].b_addr;
+		do {
+			seg_size = bvec[j].ov_vec.v_count[i];
+			if (dir)
+				memcpy(&buf_data[i * seg_size],
+				       m0_bufvec_cursor_addr(&cursor),
+				       seg_size);
+			else
+				memcpy(m0_bufvec_cursor_addr(&cursor),
+				       &buf_data[i * seg_size],
+				       seg_size);
+			++i;
+			step = m0_bufvec_cursor_step(&cursor);
+		} while (!m0_bufvec_cursor_move(&cursor, step));
+	}
+
+	M0_LEAVE();
+}
+#endif /* ISAL_ENCODE_ENABLED */
+
+#if ISAL_ENCODE_ENABLED
+static int ir_recover(struct m0_sns_ir *ir, struct m0_sns_ir_block *alive_block)
+{
+	struct m0_sns_ir_block	*failed_block;
+	struct m0_bufvec       	*alive_bufvec;
+	struct m0_bufvec       	*failed_bufvec;
+	struct m0_buf 		 buf;
+	uint8_t			 curr_idx = UINT8_MAX;
+	uint32_t 		 i;
+	uint8_t		       **failed_bufs;
+	uint32_t		 length;
+	int			 ret = 0;
+
+	M0_ENTRY("ir=%p, alive_block=%p", ir, alive_block);
+
+	/* Check if given alive block is dependecy of any failed block */
+	for (i = 0; i < ir->si_failed_nr; i++) {
+		failed_block = &ir->si_blocks[ir->si_failed_idx[i]];
+		if (m0_bitmap_get(&failed_block->sib_bitmap, alive_block->sib_idx) == 0)
+			return M0_RC(ret);
+	}
+
+	alive_bufvec = alive_block->sib_addr;
+	length = alive_bufvec->ov_vec.v_count[0] * alive_bufvec->ov_vec.v_nr;
+
+	ALLOC_ARR_INFO(failed_bufs, ir->si_failed_nr,
+		       "failed buffer array", ret);
+	if (failed_bufs == NULL)
+		goto exit;
+
+	for (i = 0; i < ir->si_failed_nr; i++) {
+		ALLOC_ARR_INFO(failed_bufs[i], length,
+			       "failed buffer", ret);
+		if (failed_bufs[i] == NULL)
+			goto exit;
+
+		m0_buf_init(&buf, failed_bufs[i], length);
+
+		failed_bufvec = ir->si_blocks[ir->si_failed_idx[i]].sib_addr;
+		/* Get data from failed vector in buffer */
+		bufvec_buf(failed_bufvec, &buf, 1, true);
+	}
+
+	for (i = 0; i < ir->si_alive_nr; i++) {
+		if(ir->si_alive_idx[i] == alive_block->sib_idx) {
+			curr_idx = i;
+			break;
+		}
+	}
+
+	buf = M0_BUF_INIT0;
+	ret = m0_buf_alloc(&buf, length);
+	if (ret != 0){
+		ret = M0_ERR_INFO(ret, "Failed to allocate buffer of "
+				  "length = %u", length);
+		goto exit;
+	}
+
+	/* Get data from current vector in buffer */
+	bufvec_buf(alive_bufvec, &buf, 1, true);
+
+	ec_encode_data_update(buf.b_nob, ir->si_data_nr, ir->si_failed_nr, curr_idx,
+			      ir->si_decode_tbls, (uint8_t *)buf.b_addr, failed_bufs);
+	m0_buf_free(&buf);
+
+	/* Copy recovered data back to buffer vector. */
+	for (i = 0; i < ir->si_failed_nr; i++) {
+		m0_buf_init(&buf, failed_bufs[i], length);
+
+		failed_block = &ir->si_blocks[ir->si_failed_idx[i]];
+		bufvec_buf(failed_block->sib_addr, &buf, 1, false);
+	}
+
+exit:
+	for (i = 0; i < ir->si_failed_nr; i++)
+		m0_free(failed_bufs[i]);
```

Comment:
If the code breaks from Line 1891, the "failed_bufs" is NULL. This will cause panic.

Reply:
Addressed.